### PR TITLE
support Huggingface's adamw

### PR DESCRIPTION
**Pull request by Tixxx**
**Description**: 
added additional mode for adamw optimizer to implement Huggingface's adamw logic

**Motivation and Context**
- matching Huggingface's adamw

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
Please add training label for future PRs.

**Comment by SherlockNoMad** [comment]
Also need to update pytroch frontend... see WeiSheng's PR for reference

**Comment by Tixxx** [comment]
> Also need to update pytroch frontend... see WeiSheng's PR for reference

yea i think his addition of the int attribute map covers my case as well, so in the map_optimizer_attribute function provided by users, they need to have "weight_decay_mode" : 0/1

File: orttraining/orttraining/core/graph/optimizer/adam_optimizer_builder.h [hunk 1]
```diff
@@ -11,7 +11,12 @@ namespace training {
 class AdamOptimizerBuilder final : public OptimizerBuilder {
  public:
   AdamOptimizerBuilder() : OptimizerBuilder("AdamOptimizer",
-                                            {"alpha", "beta", "lambda", "epsilon", "do_bias_correction"}) {}
+                                            {"alpha",
+                                             "beta",
+                                             "lambda",
+                                             "epsilon",
+                                             "do_bias_correction",
+                                             "weight_update_mode"}) {}
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/optimizer/adam_optimizer_builder.h]: >weight_update_mode [](start = 46, length = 18)

weight_update_mode or weight_decay_mode?

Comment (Tixxx) [reply orttraining/orttraining/core/graph/optimizer/adam_optimizer_builder.h]: weight_decay_mode, was a typo. the new iteration has the correct name

**Comment by SherlockNoMad** [comment]
> TEST(OptimizerTest, AdamOptimizerTest_Gradient) {

Consider move all optimizer tests to a separated optimizier_ops_test.cc

This file is growing too baig 

---
Refers to: orttraining/orttraining/test/gradient/gradient_ops_test.cc:1205 in 2695d88. [](commit_id = 2695d884e41bdee2b8992e71165317feb8ebe9da, deletion_comment = False)

File: orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h [hunk 2]
```diff
@@ -44,7 +45,8 @@ class AdamOptimizer final : public OpKernel {
   float beta_;
   float lambda_;
   float epsilon_;
-  bool do_bias_correction_;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h]: >bool [](start = 2, length = 4)

keep do_bias_correction_ as bool

File: orttraining/orttraining/training_ops/cuda/optimizer/adam.cu [hunk 3]
```diff
@@ -5,7 +5,7 @@
 #include "core/providers/cuda/cu_inc/common.cuh"
 #include "orttraining/training_ops/cuda/optimizer/common.cuh"
 #include "adam.h"
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/adam.cu]: >adam.h" [](start = 10, length = 7)

full path

File: orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h (hunk 2 above)
Comment (Tixxx) [reply orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h]: yes..

File: orttraining/orttraining/training_ops/cuda/optimizer/adam.cu [hunk 4]
```diff
@@ -69,6 +70,66 @@ __global__ void _AdamOptimizer(
...
+
+  // Apply bias correction terms on learning rate
+  const T4 step_size = T4(*eta) * _Sqrt(beta_correction) / alpha_correction;
+
+  // Huggingface updates weights in the following logic:
+  // param' = param - step_size * m1o / denom
+  // param_out = param' - original_lr * lambda * param'
+  // then param_out = param - step_size * m1o / denom - original_lr * lambda * (param - step_size * m1o / denom)
+  // so delta = -step_size * m1o / denom - original_lr * lambda * (param - step_size * m1o / denom)
+  const T4 delta = -step_size * m1o / denom - T4(*eta) * lambda * (T4(weights[id]) - step_size * m1o / denom);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/adam.cu]: >; [](start = 109, length = 1)

  // Compute the new gradient.
  if (grads_out) {
    grads_out[id] = T_GRAD(delta);
  }

still need this for parasail

File: orttraining/orttraining/training_ops/cpu/optimizer/optimizers.cc [hunk 5]
```diff
@@ -70,20 +70,52 @@ Status AdamOptimizer<T>::Compute(OpKernelContext* ctx) const {
...
+  else if (weight_decay_mode_ == 1) {
+    const auto& denom = MakeEigenArrayMap<T>(NM2).sqrt() + epsilon_;
+    const auto& step_size = eta * std::sqrt(beta_correction) / alpha_correction;
+
+    // Huggingface updates weights in the following logic:
+    // param' = param - step_size * m1o / denom
+    // param_out = param' - original_lr * lambda * param'
+    // then param_out = param - step_size * m1o / denom - original_lr * lambda * (param - step_size * m1o / denom)
+    // so delta = -step_size * m1o / denom - original_lr * lambda * (param - step_size * m1o / denom)
+    const auto& delta = -step_size * MakeEigenArrayMap<T>(NM1) / denom - eta * lambda_ * (MakeEigenArrayMap<T>(W) - step_size * MakeEigenArrayMap<T>(NM1) / denom);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/optimizer/optimizers.cc]: >const auto& delta = -step_size * MakeEigenArrayMap<T>(NM1) / denom - eta * lambda_ * (MakeEigenArrayMap<T>(W) - step_size * MakeEigenArrayMap<T>(NM1) / denom); [](start = 4, length = 159)

nit. line too long

File: orttraining/orttraining/training_ops/cuda/optimizer/adam.cu (hunk 4 above)
Comment (Tixxx) [reply orttraining/orttraining/training_ops/cuda/optimizer/adam.cu]: thanks for catching this.. deleted by mistake.

**Pull request merged**